atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,445 @@
+		Time:   time.Now().Unix(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Release the containers using in this build, so that they can be GC-ed.
+	_, err = psql.Update("containers").
+		Set("in_memory_check_build_id", nil).
+		Where(sq.Eq{"in_memory_check_build_id": b.id}).
+		RunWith(tx).
+		Exec()
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+
+	err = b.conn.Bus().Notify(buildEventsChannel(b.id))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *inMemoryCheckBuild) saveEvent(tx Tx, event atc.Event) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	payload, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+
+	_, err = psql.Insert("check_build_events").
+		Columns("event_id", "build_id", "type", "version", "payload").
+		Values(b.eventIdSeq, b.id, string(event.EventType()), string(event.Version()), payload).
+		RunWith(tx).
+		Exec()
+	b.eventIdSeq++
+
+	return err
+}
+
+func (b *inMemoryCheckBuild) Variables(logger lager.Logger, secrets creds.Secrets, varSourcePool creds.VarSourcePool) (vars.Variables, error) {
+	pipeline, found, err := b.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to find pipeline: %w", err)
+	}
+	if !found {
+		return nil, errors.New("pipeline not found")
+	}
+
+	return pipeline.Variables(logger, secrets, varSourcePool)
+}
+
+func (b *inMemoryCheckBuild) SaveEvent(ev atc.Event) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	if ev.EventType() == event.EventTypeSelectedWorker {
+		b.reallyRun = true
+	}
+
+	if !b.reallyRun {
+		b.cacheEvents = append(b.cacheEvents, ev)
+		return nil
+	}
+
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	if !b.dbInited {
+		err := b.initDbStuff(tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = b.saveEvent(tx, ev)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return b.conn.Bus().Notify(buildEventsChannel(b.id))
+}
+
+func (b *inMemoryCheckBuild) Events(from uint) (EventSource, error) {
+	notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsChannel(b.id), func() (bool, error) {
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newBuildEventSource(
+		b.id,
+		"check_build_events",
+		b.conn,
+		notifier,
+		from,
+		true,
+	), nil
+}
+
+// AbortNotifier returns NoopNotifier because there is no way to abort a in-memory
+// check build. Say a in-memory build may run on ATC-a, but abort-build API call
+// might be received by ATC-b, there is not a channel for ATC-b to tell ATC-a to
+// mark the in-memory build as aborted. If we really want to abort a in-memory
+// check build in future, it might need to add a new table "aborted-in-memory-builds"
+// and API insert in-memory build id to the table, and AbortNotifier watches the
+// table to see if current build should be aborted.
```

Comment:
Doesn't make a huge difference since we aren't implementing it now, but in case someone comes along looking to implement this, maybe we could point them in a slightly different direction. Rather than a table, using the notification bus probably makes more sense here - so maybe we could suggest that instead

Reply:
If I understand correctly, a notification bus needs to watch if a build is aborted or not. For normal build, notification bus `select aborted from builds where id=xx`. For in memory build, there is no way for notification bus to watch if a build is aborted, that reason why I suggested a new table. Please correct me.